File: backend/src/services/steganography.js
import { spawn } from 'child_process';
import { writeFileSync, readFileSync, unlinkSync, existsSync } from 'fs';
import { join } from 'path';
import { randomBytes } from 'crypto';
import sharp from 'sharp';

/**
 * Steganography service using OpenCV Python scripts
 * Embeds encrypted documents into carrier images
 */
class SteganographyService {
  constructor() {
    this.tempDir = join(process.cwd(), 'temp');
    this.pythonScriptPath = join(process.cwd(), 'src', 'python');
    this.ensureTempDir();
    this.ensurePythonScripts();
  }

  // Ensure temp directory exists
  ensureTempDir() {
    if (!existsSync(this.tempDir)) {
      import('fs').then(fs => fs.mkdirSync(this.tempDir, { recursive: true }));
    }
  }

  // Create Python scripts for steganography
  ensurePythonScripts() {
    if (!existsSync(this.pythonScriptPath)) {
      import('fs').then(fs => fs.mkdirSync(this.pythonScriptPath, { recursive: true }));
    }

    // Create embed script
    const embedScript = `
import cv2
import numpy as np
import sys
import json
import base64

def embed_data_in_image(image_path, data, output_path):
    """Embed binary data into an image using LSB steganography"""
    try:
        # Read the carrier image
        img = cv2.imread(image_path)
        if img is None:
            raise ValueError("Could not read image file")
        
        # Convert data to binary
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        # Convert to binary string
        binary_data = ''.join(format(byte, '08b') for byte in data)
        
        # Add delimiter to mark end of data
        binary_data += '1111111111111110'  # 16-bit delimiter
        
        # Check if image can hold the data
        total_pixels = img.shape[0] * img.shape[1] * img.shape[2]
        if len(binary_data) > total_pixels:
            raise ValueError("Image too small to hold the data")
        
        # Flatten image
        flat_img = img.flatten()
        
        # Embed data using LSB
        for i, bit in enumerate(binary_data):
            flat_img[i] = (flat_img[i] & 0xFE) | int(bit)
        
        # Reshape back to original shape
        stego_img = flat_img.reshape(img.shape)
        
        # Save the steganographic image
        success = cv2.imwrite(output_path, stego_img)
        if not success:
            raise ValueError("Could not save steganographic image")
        
        return True
        
    except Exception as e:
        print(f"Error in embed_data_in_image: {str(e)}", file=sys.stderr)
        return False

if __name__ == "__main__":
    if len(sys.argv) != 4:
        print("Usage: python embed.py <image_path> <data_file> <output_path>", file=sys.stderr)
        sys.exit(1)
    
    image_path = sys.argv[1]
    data_file = sys.argv[2]
    output_path = sys.argv[3]
    
    try:
        # Read data from file
        with open(data_file, 'rb') as f:
            data = f.read()
        
        # Embed data
        if embed_data_in_image(image_path, data, output_path):
            print(json.dumps({"success": True, "message": "Data embedded successfully"}))
        else:
            print(json.dumps({"success": False, "error": "Failed to embed data"}))
            sys.exit(1)
            
    except Exception as e:
        print(json.dumps({"success": False, "error": str(e)}), file=sys.stderr)
        sys.exit(1)
`;

    // Create extract script
    const extractScript = `
import cv2
import numpy as np
import sys
import json

def extract_data_from_image(image_path):
    """Extract binary data from steganographic image using LSB"""
    try:
        # Read the steganographic image
        img = cv2.imread(image_path)
        if img is None:
            raise ValueError("Could not read image file")
        
        # Flatten image
        flat_img = img.flatten()
        
        # Extract LSBs
        binary_data = ""
        delimiter = '1111111111111110'
        
        for pixel in flat_img:
            binary_data += str(pixel & 1)
            
            # Check for delimiter
            if binary_data.endswith(delimiter):
                binary_data = binary_data[:-len(delimiter)]
                break
        
        # Convert binary to bytes
        if len(binary_data) % 8 != 0:
            # Pad with zeros if necessary
            binary_data += '0' * (8 - len(binary_data) % 8)
        
        # Convert to bytes
        data = bytearray()
        for i in range(0, len(binary_data), 8):
            byte = binary_data[i:i+8]
            if len(byte) == 8:
                data.append(int(byte, 2))
        
        return bytes(data)
        
    except Exception as e:
        print(f"Error in extract_data_from_image: {str(e)}", file=sys.stderr)
        return None

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python extract.py <image_path> <output_file>", file=sys.stderr)
        sys.exit(1)
    
    image_path = sys.argv[1]
    output_file = sys.argv[2]
    
    try:
        # Extract data
        data = extract_data_from_image(image_path)
        
        if data is not None:
            # Save extracted data
            with open(output_file, 'wb') as f:
                f.write(data)
            
            print(json.dumps({"success": True, "message": "Data extracted successfully"}))
        else:
            print(json.dumps({"success": False, "error": "Failed to extract data"}))
            sys.exit(1)
            
    except Exception as e:
        print(json.dumps({"success": False, "error": str(e)}), file=sys.stderr)
        sys.exit(1)
`;

    // Write Python scripts
    writeFileSync(join(this.pythonScriptPath, 'embed.py'), embedScript);
    writeFileSync(join(this.pythonScriptPath, 'extract.py'), extractScript);
  }

  // Generate a suitable carrier image if none provided
  async generateCarrierImage(width = 1024, height = 768) {
    try {
      const filename = `carrier_${randomBytes(8).toString('hex')}.png`;
      const filepath = join(this.tempDir, filename);

      // Create a random noise image using Sharp
      const buffer = Buffer.alloc(width * height * 3);
      for (let i = 0; i < buffer.length; i++) {
        buffer[i] = Math.floor(Math.random() * 256);
      }

      await sharp(buffer, {
        raw: {
          width: width,
          height: height,
          channels: 3
        }
      })
      .png()
      .toFile(filepath);

      return filepath;
    } catch (error) {
      console.error('Error generating carrier image:', error);
      throw new Error('Failed to generate carrier image');
    }
  }

  // Embed encrypted document into image
  async embedDocument(encryptedData, carrierImagePath = null) {
    const tempId = randomBytes(8).toString('hex');
    const dataFile = join(this.tempDir, `data_${tempId}.bin`);
    const outputFile = join(this.tempDir, `stego_${tempId}.png`);
    
    let actualCarrierPath = carrierImagePath;

    try {
      // Generate carrier image if not provided
      if (!actualCarrierPath) {
        actualCarrierPath = await this.generateCarrierImage();
      }

      // Write encrypted data to temporary file
      writeFileSync(dataFile, encryptedData);

      // Run Python embedding script
      const result = await this.runPythonScript('embed.py', [
        actualCarrierPath,
        dataFile,
        outputFile
      ]);

      if (!result.success) {
        throw new Error(result.error || 'Embedding failed');
      }

      // Read the steganographic image
      const stegoImageBuffer = readFileSync(outputFile);

      // Cleanup temporary files
      this.cleanup([dataFile, outputFile]);
      if (!carrierImagePath) {
        this.cleanup([actualCarrierPath]);
      }

      return {
        success: true,
        imageBuffer: stegoImageBuffer,
        size: stegoImageBuffer.length
      };

    } catch (error) {
      console.error('Embedding error:', error);
      
      // Cleanup on error
      this.cleanup([dataFile, outputFile]);
      if (!carrierImagePath) {
        this.cleanup([actualCarrierPath]);
      }

      return {
        success: false,
        error: error.message
      };
    }
  }

  // Extract encrypted document from steganographic image
  async extractDocument(stegoImageBuffer) {
    const tempId = randomBytes(8).toString('hex');
    const imageFile = join(this.tempDir, `stego_${tempId}.png`);
    const outputFile = join(this.tempDir, `extracted_${tempId}.bin`);

    try {
      // Write steganographic image to temporary file
      writeFileSync(imageFile, stegoImageBuffer);

      // Run Python extraction script
      const result = await this.runPythonScript('extract.py', [
        imageFile,
        outputFile
      ]);

      if (!result.success) {
        throw new Error(result.error || 'Extraction failed');
      }

      // Read extracted data
      const extractedData = readFileSync(outputFile);

      // Cleanup temporary files
      this.cleanup([imageFile, outputFile]);

      return {
        success: true,
        data: extractedData
      };

    } catch (error) {
      console.error('Extraction error:', error);
      
      // Cleanup on error
      this.cleanup([imageFile, outputFile]);

      return {
        success: false,
        error: error.message
      };
    }
  }

  // Run Python script with arguments
  async runPythonScript(scriptName, args = []) {
    return new Promise((resolve, reject) => {
      const scriptPath = join(this.pythonScriptPath, scriptName);
      const pythonProcess = spawn('python', [scriptPath, ...args]);

      let stdout = '';
      let stderr = '';

      pythonProcess.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      pythonProcess.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      pythonProcess.on('close', (code) => {
        if (code === 0) {
          try {
            const result = JSON.parse(stdout.trim());
            resolve(result);
          } catch (error) {
            resolve({ success: true, message: stdout.trim() });
          }
        } else {
          try {
            const error = JSON.parse(stderr.trim());
            resolve(error);
          } catch (parseError) {
            resolve({ 
              success: false, 
              error: stderr.trim() || `Process exited with code ${code}` 
            });
          }
        }
      });

      pythonProcess.on('error', (error) => {
        resolve({
          success: false,
          error: `Failed to start Python process: ${error.message}`
        });
      });
    });
  }

  // Cleanup temporary files
  cleanup(files) {
    files.forEach(file => {
      try {
        if (existsSync(file)) {
          unlinkSync(file);
        }
      } catch (error) {
        console.warn(`Failed to cleanup file ${file}:`, error.message);
      }
    });
  }
}

export default new SteganographyService();

File: src/services/steganography.py
#!/usr/bin/env python3
"""
Steganography service for hiding and extracting documents in images
Uses LSB (Least Significant Bit) technique with OpenCV
"""

import cv2
import numpy as np
import sys
import json
import base64
import os
from typing import Tuple, Optional

class SteganographyService:
    def __init__(self):
        self.delimiter = "###END_OF_DATA###"
        self.header_size = 32  # Size for storing data length
    
    def hide_data_in_image(self, image_path: str, data: bytes, output_path: str) -> bool:
        """
        Hide binary data in an image using LSB steganography
        
        Args:
            image_path: Path to carrier image
            data: Binary data to hide
            output_path: Path for output steganographic image
            
        Returns:
            bool: Success status
        """
        try:
            # Read the carrier image
            image = cv2.imread(image_path)
            if image is None:
                raise ValueError(f"Could not read image: {image_path}")
            
            # Convert data to binary string
            data_with_delimiter = data + self.delimiter.encode()
            binary_data = ''.join(format(byte, '08b') for byte in data_with_delimiter)
            data_length = len(binary_data)
            
            # Check if image can hold the data
            max_capacity = image.shape[0] * image.shape[1] * image.shape[2]
            if data_length > max_capacity:
                raise ValueError(f"Image too small to hold data. Need {data_length} bits, have {max_capacity}")
            
            # Flatten image for easier processing
            flat_image = image.flatten()
            
            # Hide data length in first 32 bits
            length_binary = format(data_length, '032b')
            for i in range(32):
                flat_image[i] = (flat_image[i] & 0xFE) | int(length_binary[i])
            
            # Hide actual data
            for i, bit in enumerate(binary_data):
                pixel_index = i + 32
                flat_image[pixel_index] = (flat_image[pixel_index] & 0xFE) | int(bit)
            
            # Reshape back to original image shape
            stego_image = flat_image.reshape(image.shape)
            
            # Save steganographic image
            success = cv2.imwrite(output_path, stego_image)
            if not success:
                raise ValueError(f"Could not save image: {output_path}")
            
            return True
            
        except Exception as e:
            print(f"Error hiding data: {str(e)}", file=sys.stderr)
            return False
    
    def extract_data_from_image(self, image_path: str) -> Optional[bytes]:
        """
        Extract hidden data from steganographic image
        
        Args:
            image_path: Path to steganographic image
            
        Returns:
            bytes: Extracted data or None if failed
        """
        try:
            # Read the steganographic image
            image = cv2.imread(image_path)
            if image is None:
                raise ValueError(f"Could not read image: {image_path}")
            
            # Flatten image
            flat_image = image.flatten()
            
            # Extract data length from first 32 bits
            length_binary = ''
            for i in range(32):
                length_binary += str(flat_image[i] & 1)
            
            data_length = int(length_binary, 2)
            
            # Validate data length
            if data_length <= 0 or data_length > len(flat_image) - 32:
                raise ValueError("Invalid data length extracted")
            
            # Extract binary data
            binary_data = ''
            for i in range(data_length):
                pixel_index = i + 32
                binary_data += str(flat_image[pixel_index] & 1)
            
            # Convert binary to bytes
            extracted_bytes = bytearray()
            for i in range(0, len(binary_data), 8):
                byte_chunk = binary_data[i:i+8]
                if len(byte_chunk) == 8:
                    extracted_bytes.append(int(byte_chunk, 2))
            
            # Find delimiter and extract actual data
            extracted_data = bytes(extracted_bytes)
            delimiter_bytes = self.delimiter.encode()
            
            if delimiter_bytes in extracted_data:
                actual_data = extracted_data.split(delimiter_bytes)[0]
                return actual_data
            else:
                raise ValueError("Delimiter not found in extracted data")
                
        except Exception as e:
            print(f"Error extracting data: {str(e)}", file=sys.stderr)
            return None
    
    def create_carrier_image(self, width: int = 1920, height: int = 1080) -> np.ndarray:
        """
        Create a random carrier image for steganography
        
        Args:
            width: Image width
            height: Image height
            
        Returns:
            numpy.ndarray: Generated carrier image
        """
        # Create random noise image
        image = np.random.randint(0, 256, (height, width, 3), dtype=np.uint8)
        
        # Add some structure to make it look more natural
        # Add gradient
        for i in range(height):
            for j in range(width):
                image[i, j] = image[i, j] * (0.7 + 0.3 * (i + j) / (height + width))
        
        # Apply Gaussian blur for smoother appearance
        image = cv2.GaussianBlur(image, (5, 5), 0)
        
        return image.astype(np.uint8)

def main():
    """
    Command line interface for steganography operations
    """
    if len(sys.argv) < 2:
        print("Usage: python steganography.py <command> [args...]")
        print("Commands:")
        print("  hide <image_path> <data_base64> <output_path>")
        print("  extract <image_path>")
        print("  create_carrier <output_path> [width] [height]")
        sys.exit(1)
    
    stego = SteganographyService()
    command = sys.argv[1]
    
    try:
        if command == "hide":
            if len(sys.argv) != 5:
                raise ValueError("hide command requires: image_path data_base64 output_path")
            
            image_path = sys.argv[2]
            data_base64 = sys.argv[3]
            output_path = sys.argv[4]
            
            # Decode base64 data
            data = base64.b64decode(data_base64)
            
            success = stego.hide_data_in_image(image_path, data, output_path)
            result = {"success": success, "output_path": output_path if success else None}
            print(json.dumps(result))
            
        elif command == "extract":
            if len(sys.argv) != 3:
                raise ValueError("extract command requires: image_path")
            
            image_path = sys.argv[2]
            extracted_data = stego.extract_data_from_image(image_path)
            
            if extracted_data:
                # Encode to base64 for safe transport
                data_base64 = base64.b64encode(extracted_data).decode('utf-8')
                result = {"success": True, "data": data_base64}
            else:
                result = {"success": False, "data": None}
            
            print(json.dumps(result))
            
        elif command == "create_carrier":
            if len(sys.argv) < 3:
                raise ValueError("create_carrier command requires: output_path [width] [height]")
            
            output_path = sys.argv[2]
            width = int(sys.argv[3]) if len(sys.argv) > 3 else 1920
            height = int(sys.argv[4]) if len(sys.argv) > 4 else 1080
            
            carrier_image = stego.create_carrier_image(width, height)
            success = cv2.imwrite(output_path, carrier_image)
            
            result = {"success": success, "output_path": output_path if success else None}
            print(json.dumps(result))
            
        else:
            raise ValueError(f"Unknown command: {command}")
            
    except Exception as e:
        error_result = {"success": False, "error": str(e)}
        print(json.dumps(error_result))
        sys.exit(1)

if __name__ == "__main__":
    main()